### PR TITLE
[Muti_backend]part_2_device

### PR DESCRIPTION
**Pull request by Galaxy1458**
### PR Category
Other

### Type of Change
New Feature 

### Description
muti_backend part_2，This PR solves the problem of uniform device use，Replace all field codes with 'cuda' with variables such as device，e.g.：
```python
#demo_1
torch.randn((128, 128, 128), device=flag_gems.device)

#demo_2
with torch_backend.device(x.device)

#demo_3
from flag_gems.utils.random_utils import philox_backend_seed_offset

#demo_4
from ..runtime.moduel_tool import tl_extra_module
rsqrt = tl_extra_module.rsqrt

......
```

### Progress

- [x] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/runtime/register.py [hunk 1]
```diff
@@ -81,7 +79,7 @@ def get_forward_ops(self) -> list[str]:
         return self.forward_ops if self.debug else []
```
Comment (StrongSpoon) [inline src/flag_gems/runtime/register.py:79]: python 3.8 doesn't support expression like `-> list[str]`.

File: README.md [hunk 2]
```diff
@@ -138,16 +138,16 @@ pip install .
...
+    B = torch.randn((K, N), dtype=torch.float16, device=flag_gems.device)
     with flag_gems.use_gems():
         C = torch.mm(A, B)
     ```
 
 ### Execute
 
 1. Test Operator Accuracy
-    - Run reference on cuda
+    - Run reference on flag_gems.device
```
Comment (StrongSpoon) [inline README.md]: change the corresponding part of Chinese readme.

File: benchmark/performance_utils.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 import yaml
 
 import flag_gems
+from flag_gems.runtime import torch_backend, torch_backends_device_fn
```
Comment (StrongSpoon) [inline benchmark/performance_utils.py]: names torch_backend and torch_backends_device_fn are confusing.

File: src/flag_gems/ops/arange.py [hunk 4]
```diff
@@ -42,8 +43,8 @@ def arange_start(
         pin_memory = False
 
     if device is None:
-        device = torch.device(
-            "cuda"
+        device = (
+            runtime.device.name
```
Comment (StrongSpoon) [inline src/flag_gems/ops/arange.py:47]: what's the type of runtime.device.name? is it a string?

File: src/flag_gems/ops/gather.py [hunk 5]
```diff
@@ -14,6 +14,7 @@ def generate_imports(code: IndentedBuffer) -> IndentedBuffer:
     code.writeline("import torch")
     code.writeline("import triton")
     code.writeline("import triton.language as tl")
+    code.writeline("from flag_gems.runtime import torch_backend")
```
Comment (StrongSpoon) [inline src/flag_gems/ops/gather.py]: there is no place using torch_backend. what about adding device guard for it?

File: src/flag_gems/ops/gelu.py [hunk 6]
```diff
@@ -4,15 +4,13 @@
...
 from ..utils import pointwise_dynamic
 
-try:
-    from triton.language.extra.cuda.libdevice import erf, exp, pow, tanh
-except ImportError:
-    try:
-        from triton.language.math import erf, exp, pow, tanh
-    except ImportError:
-        from triton.language.libdevice import erf, exp, pow, tanh
+erf = tl_extra_module.erf
```
Comment (StrongSpoon) [inline src/flag_gems/ops/gelu.py:10]: from tl_extra_module import era, exp, pow, tanh

File: src/flag_gems/runtime/backend/__init__.py [hunk 7]
```diff
@@ -33,18 +36,37 @@ def gen_torch_tensor_attr_res(tensor, attr_name):
...
+    global device_name, tl_extra_backend_module
+    if tl_extra_backend_module is not None:
+        return tl_extra_backend_module
+    device_name = device_name or get_vendor_info(vendor_name).device_name
+    module_str = f"triton.language.extra.{device_name}.libdevice"
+    tl_extra_backend_module = importlib.import_module(module_str)
+    return tl_extra_backend_module
+
+
+def gen_torch_backends_device(vendor_name=None):
```
Comment (StrongSpoon) [inline src/flag_gems/runtime/backend/__init__.py]: once flag_gems is imported, gen_torch_backends_device is called in runtime.__init__. why do we need to check if torch_backends_device is none?

File: src/flag_gems/runtime/backend/__init__.py [hunk 8]
```diff
@@ -33,18 +36,37 @@ def gen_torch_tensor_attr_res(tensor, attr_name):
...
+    global device_name, torch_backends_device
+    if torch_backends_device is not None:
+        return torch_backends_device
+    device_name = device_name or get_vendor_info(vendor_name).device_name
+    module_str = f"torch.backends.{device_name}"
+    torch_backends_device = importlib.import_module(module_str)
+    return torch_backends_device
+
+
+def gen_torch_device_object(vendor_name=None):
```
Comment (StrongSpoon) [inline src/flag_gems/runtime/backend/__init__.py:63]: ditto

File: tests/test_named_ops.py [hunk 9]
```diff
@@ -3,6 +3,9 @@
 
 import pytest
 
+import flag_gems
```
Comment (StrongSpoon) [inline tests/test_named_ops.py:6]: just delete this file. it's useless now on.

**Review by StrongSpoon — COMMENTED** [review]
done:)

File: benchmark/performance_utils.py (hunk 3 above)
Comment (Galaxy1458) [reply benchmark/performance_utils.py]: torch_backend is like 'torch.cuda' object and torch_backends_device_fn is like 'torch.backend.cuda' object. annotations are also added to 'flag_gem/runtime/__init__.py'

File: src/flag_gems/ops/arange.py (hunk 4 above)
Comment (Galaxy1458) [reply src/flag_gems/ops/arange.py:47]: yes

File: src/flag_gems/ops/gather.py (hunk 5 above)
Comment (Galaxy1458) [reply src/flag_gems/ops/gather.py]: removed

File: src/flag_gems/ops/gelu.py (hunk 6 above)
Comment (Galaxy1458) [reply src/flag_gems/ops/gelu.py:10]: Syntax such as ' from xxx import 'is illegal for libraries imported using the 'importlib' library

File: src/flag_gems/runtime/backend/__init__.py (hunk 7 above)
Comment (Galaxy1458) [reply src/flag_gems/runtime/backend/__init__.py]: Although currently not checking for None is not problematic, if functionality is added later, this method might need to be called multiple times（for convenience. This is also a conventional practice.

File: src/flag_gems/runtime/backend/__init__.py (hunk 8 above)
Comment (Galaxy1458) [reply src/flag_gems/runtime/backend/__init__.py:63]: ditto

File: src/flag_gems/ops/arange.py (hunk 4 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/arange.py:47]: divide before modification is a instance of torch.device. it doesn't cause error though. I suggest deleting the brackets.

File: src/flag_gems/ops/gelu.py (hunk 6 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/gelu.py:10]: got it

File: src/flag_gems/runtime/backend/__init__.py (hunk 7 above)
Comment (StrongSpoon) [reply src/flag_gems/runtime/backend/__init__.py]: will torch_backends_device change in one process?

File: tests/test_named_ops.py [hunk 10]
```diff
@@ -1,242 +0,0 @@
-import argparse
```
Comment (StrongSpoon) [inline tests/test_named_ops.py]: FlagPerf needs this file still. please redo it. I'm sorry orz.

**Review by StrongSpoon — APPROVED** [review]
lg

**Review by StrongSpoon — APPROVED** [review]
lgtm
